src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs

```diff
@@ -74,24 +74,108 @@ private static Delegate CreateDelegateProxy(Type typeOfProxy, MethodInfo invokeM
         }
 
         private static Expression CreateBodyExpression(
-            MethodInfo delegateMethod, DelegateCallInterceptedEventRaiser eventRaiser, IEnumerable<Expression> parameterExpressions)
+            MethodInfo delegateMethod,
+            DelegateCallInterceptedEventRaiser eventRaiser,
+            ParameterExpression[] parameterExpressions)
         {
-            var parameterExpressionsCastToObject =
-                parameterExpressions.Select(x => Expression.Convert(x, typeof(object))).Cast<Expression>().ToArray();
-
-            Expression body = Expression.Call(
+            /*
+             * Generate a method that:
+             * - wraps its arguments in an object array
+             * - passes this array to eventRaiser.Raise()
+             * - assigns the output values back to the ref/out parameters
+             * - casts and returns the result of eventRaiser.Raise()
+             *
+             * For instance, for a delegate like this:
+             *
+             * delegate int Foo(int x, ref int y, out int z);
+             *
+             * We generate a method like this:
+             *
+             * int ProxyFoo(int x, ref int y, out int z)
+             * {
+             *     var arguments = new[]{ (object)x, (object)y, (object)z };
+             *     var result = (int)eventRaiser.Raise(arguments);
+             *     y = (int)arguments[1];
+             *     z = (int)arguments[2];
+             *     return result;
+             * }
+             *
+             * Or, for a delegate with void return type:
+             *
+             * delegate void Foo(int x, ref int y, out int z);
+             *
+             * void ProxyFoo(int x, ref int y, out int z)
+             * {
+             *     var arguments = new[]{ (object)x, (object)y, (object)z };
+             *     eventRaiser.Raise(arguments);
+             *     y = (int)arguments[1];
+             *     z = (int)arguments[2];
+             * }
+             */

```

Comment:
@adamralph, see the first comment thread (hidden - click "show comments" where it says "blairconrad and 1 other commented on an outdated diff") for the discussion about this.

Basically, if you have `//` comments followed by a blank line, StyleCop complains (because obviously, it knows better than us mere humans...). In the case of `Should_track_constraints_supplied_in_calls_made_from_overlapping_threads`, it works only because there's another comment after the blank line - if you remove it, it gives SA1512. That's why I used `/* */`. The alternatives are:

- use `////` instead of `//` (looks a bit weird)
- don't leave a blank line after the comment block (but it looks like it only applies to the next line)
- put another comment before the next line of code
- suppress the warning
- drop StyleCop :stuck_out_tongue: